src/textual/_work_decorator.py

```diff
@@ -0,0 +1,102 @@
+from __future__ import annotations
+
+from functools import partial, wraps
+from typing import TYPE_CHECKING, Callable, Coroutine, TypeVar, Union, cast, overload
+
+from typing_extensions import ParamSpec, TypeAlias
+
+if TYPE_CHECKING:
+    from .worker import Worker
+
+
+FactoryParamSpec = ParamSpec("FactoryParamSpec")
+DecoratorParamSpec = ParamSpec("DecoratorParamSpec")
+ReturnType = TypeVar("ReturnType")
+
+Decorator: TypeAlias = Callable[
```

Comment:
"Decorator" sounds like a very generic name that encompasses Python decorators.
Maybe something like "WorkerDecorator" would be more appropriate?